### ISSUE TITLE
How to train on Pascal dataset?

### ISSUE THREAD

**Issue by Exuro889** [shekkizh/FCN.tensorflow#16]
Hello,

I was wondering what all is need to be done to get this working with the Pascal dataset. I see that the output placeholder channel size is hardcoded to 1 and the pascal annotations are rgb images so that would need to be changed to 3, along with the number of classes and whatnot. I've tried changing those but its giving me an error in the loss function line and I'm having a hard time understanding how logits which is of shape (?, ?, ?, num_classes) can be compared with y_output which is of shape (?, width, height, channel).

Also a separate question, but do you know how to compute intersection over union for the output?

Thanks

edit: Spent a little bit of time looking around and it looks like I need to figure out how to map the color mapped segmentation labels they give us in the dataset with a 0-20 integer indexed version

**Comment by xtudbxk** [comment]
Hello, 
I also tried the code on PASCAL VOC dataset. What I did is to change the PASCAL VOC annotations rgb image to the gray level image, so I need not to change the output placeholder channel size.  Maybe this solution help you. And for the cross entropy loss, I looked throught the tensorflow web and found that:
the function name is sparse_softmax_cross_entropy_with_logits, so in the first step, the logit which origin shape is (?, weight, height, num_classes would be input to a softmax layer, its shape changes to (?, weight, height) , and the labes is tf.sqeeuze(annotations) which will change the (?, width, height, channel) to (?, width, height) when the channel is 1.
But unfortunately, I encountered the problem that the loss is Nan, And don't find a solution. 
Do you finish your experiment and solve it ?

Thanks,

**Comment by Exuro889** [comment]
Hi,

I was able to get it working, though I still have some issues. First off I also set the image to grayscale, though I relabeled the pixels to an integer class number as seen here. http://host.robots.ox.ac.uk/pascal/VOC/voc2012/segexamples/index.html

I was also having the NAN issue, but that was because the class number that I was getting for some pixels was outside the range of the classes, IE, you have 21 classes, but the void labels are labeled 255 and so it goes nan as its filling up one hot vectors of size 21, so trying to fill in teh 255th slot in a 21 slot one hot vector doesn't exactly work..

So one issue is theres this void label, which is the outline of the objects. I'm not sure how to get rid of this or ignore it. In caffe apparently it's a simple argument setup in their softmax to ignore the label. Since I don't know how to fix this I made their values 22 so I have 22 classes and my output will be width x height x 22.

With it working though I'm not getting anywhere near the results of the fully convolutional networks paper. They state they need atleast 175 epochs before state of the art results and my results dont look anything near state of the art at 250, so I'm not sure if that is an issue with the void labels. Perhaps it has to do with my input size only being 224x224. I only have an 8GB gpu and the model itself is almost maxing it out so I'm limited in that space.

If anyone knows how to ignore the void label in tensorflow I would be happy to hear about it

**Comment by Exuro889** [comment]
Well I got some results and they aren't too pretty. I removed the void label in preprocessing so that plays a little bit into it. This is a little over 200 epochs(they state 175 needed for state of the art)

accuracy: 0.7878521022151096
mean accuracy: 0.22906130845951078
mean iou: 0.17522699426670316
f.w iou: 0.6675439921050957

I looked at some validation output images and they did not look good. I don't know what the issue is. In the paper it talked about having weight decay so I'm wondering if that is playing into it. I can understand the resizing of the images being an issue as the papers are at 500 x 500, but this is just straight up not good.

**Comment by shekkizh** [comment]
@Exuro889 - if possible can you share few of the results on PASCAL VOC?!

**Comment by Exuro889** [comment]
label: ![](http://i.imgur.com/H1a3g2S.jpg) prediction: ![](http://i.imgur.com/JRMf0XC.jpg)

label: ![](http://i.imgur.com/fH58k4w.jpg) prediction: ![](http://i.imgur.com/wDQP1lM.jpg)

label: ![](http://i.imgur.com/pxQbhZU.jpg) prediction: ![](http://i.imgur.com/ZAA19Dh.jpg)

So the two that look ok are few and far between unfortunately. The MIT set actually produces really good indoor scenes in comparison. If you have any thoughts on what is wrong let me know. Right now I'm guessing the system needs more regularization, as validation loss is pretty consistent while training loss goes down. The training predictions look fantastic, so it's doing something. The training set doesn't represent the validation set well enough or something. They are split 1500 and 1500 though, which honestly I feel 1500 isn't enough, especially when the MIT dataset has 20k to 2k.

**Comment by shekkizh** [comment]
@Exuro889 Thanks for sharing your results. A few thoughts I have 
 - Given you see improvement in training but validation has plateaued would mean you are overfitting. As you guessed regularization could help you with this. Using a larger batch size might reduce overfitting as well.
- The hyper parameters, optimization routines might all contribute to having different results from that of the paper.
- The prediction for person in bike seems interesting in that the prediction is in parts. There are later papers which specifically train to avoid such problems.

**Comment by obendidi** [comment]
Hello , @Exuro889 can you possibly give more details on how you relabled the annotations to an integer class number , I'm trying to train the model on my own dataset , and I'm having the same issue of the nan value while training, thanks !

**Comment by xtudbxk** [comment]
Just label the classes in order. For example, 0 for background, 1 for person, 2 for bird, ...
The reason why the loss is nan is that the function sparse_softmax_cross_entropy_with_logits() requires the labels is an index in [0, num_of_class), which you can learn more from https://www.tensorflow.org/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits

**Comment by SoulGuedria** [comment]
@xtudbxk How can i label the classes in my own dataset please ? Is there any annotation file ? Thanks a lot :)

**Comment by Chen-Song** [comment]
Hi, @Exuro889 ,
     Have you solved this problem? I meet this problem too, and the result in pascal voc 2012 is 
val_loss: 1.436496967
accuracy:0.72672651
mean accuracy: 0.106219544
mean iou: 0.062652998
f.w iou:0.597502524
     Look forward to your reply. Thanks.

**Comment by varungupta31** [comment]
Hey @Chen-Song @Exuro889 
I'm trying to run the model on custom data as well.
Can you help with how you are able to get these evaluation metrics - accuracy, mean accuracy, mean iou, f.w iou ?

Thank you.
